mbbsd/pfterm.c

```diff
@@ -497,9 +501,19 @@ endwin(void)
 
 int
 resizeterm(int rows, int cols)
+{
+    // assume that the true terminal size does not exceed 9999x9999
+    return resizeterm_true(rows, cols, 9999, 9999);
```

Comment:
No. Just assume the limited resolution is the same as the full resolution. We won't be using the old API, but we want to keep the behavior the same.

Reply:
I doubted about which behavior to keep. The pros and cons of which I thought are listed below.

I will first implement option 1 as suggested however.

There are 3 options regarding ptt/pttbbs#120. The first 2 are incompatible with each other.

1. keep the behavior as before the forward scrolling workaround (suggested)
    * Pros:
        * It avoids the undefined effect described by [ECMA-48](https://www.ecma-international.org/wp-content/uploads/ECMA-48_5th_edition_june_1991.pdf) 6.1.7 due to out-of-bound coordinates.
        * Terminal-parsing&ndash;based 3rd-party APPs are guaranteed to work.
            * However, Ptt & Ptt2 are the only 2 sites which I know to have such 3rd-party APPs. Both sites will certainly use the new API and might make this irrelevant.
    * Con: BBS maintainers have to modify their code to resolve the forward scrolling issue.
        * This task may be trivial for PttBBS-based BBSes. Not necessarily so for non&ndash;PttBBS-based BBSes.
        * For PttBBS-based BBSes, it is possible that their codes either are too lagged behind PttBBS to work with 3rd-party APPs or are recent enough to simply use the new API by applying the modifications from PttBBS.
        * For non&ndash;PttBBS-based BBSes, however, I suspect that there are currently few than 10 people who are actively maintaining a non&ndash;PttBBS-based BBS which use pfterm. Dream-Land BBS is the only site which I know to both be such a BBS and have the site currently accessible.
2. keep the behavior as introduced in the workaround (the initial choice of the `resizeterm_true()` solution of this PR)
    * Pro: BBS maintainers need only to update `pfterm.c` to resolve the issue. Using the new API is optional.
    * Con: APP maintainers may need to update their APPs for handling out-of-bound screen coordinates.
        * At this time (2023-05-29 04:20 UTC+8), almost all such APPs which I know to have been investigated either are reported to be immune or have already updated to handle it.
3. rework `fterm_rawscroll()` to make forward scrolling work independent of the actual terminal size (out of scope)
    * Pro: It avoids changing the API and does not introduce a (`struct`ed) global variable which is only used in one place.
    * It is possible to achieve this by sending Delete Line (`"\x1b[%d;M"`) per ECMA-48 and the [VT102 User Guide](https://www.vt100.net/docs/vt102-ug/chapter5.html#S5.5.2.21).
        * However, this escape may not be supported by some terminal emulators and terminal-parsing&ndash;based 3rd-party APPs.
        * [MapleBBS 3.00a](https://github.com/bbsmirror/BBSmirror/blob/master/Maple/Maple3/Old/Alpha/MapleBBS-3.00a.tgz) and on used Insert Line (`"\x1b[%d;L"`) for backward scrolling but fell back to the move-to-bottom-then-insert-newlines method for forward scrolling, which seems to be a refactored version of the scrolling handling of [MapleBBS 2.36b](https://github.com/bbsmirror/BBSmirror/blob/master/Maple/Maple2/MapleBBS-2.36b.tar.gz) (inherited by the non-pfterm `screen.c` of PttBBS).
        * This option needs compatibility investigation, which is out of the scope of this PR.

Reply:
There's only two invocations of this function. One is even inside pfterm's init function. And you are updating them both.

If someone is exporting pfterm to another application or system, that's up to them.  Either they follow curses / ncurses API definition, which this is based on, or they switch to the new custom function. We already introduced a regression anyways. Better to revert.

Regarding 3., that's not actually scrolling the terminal, and probably leads to very ugly content in the scrollback buffer. I believe pfterm was optimized to use proper scrolling to avoid needless clearing and redrawing.

BTW, the resolution cropping thing predates git history, so there's no conclusive way to say what the reason for it was. But I guess it was to avoid users randomly resizing to an insanely huge screen size, either really using it, or sending a bogus stream, and consuming huge amounts of memory for the local terminal buffer. If so then the cropping should really live in the terminal implementations (`pfterm.c` and `screen.c`). The BBS terminal/window sizes should be updated by somehow reading back the new terminal size after resizeterm() is called. ncurses provides the global variables `COLS` and `LINES` for this. Something to think about even though out of scope.